qiskit/circuit/quantumcircuit.py

```diff
@@ -4067,11 +4071,19 @@ def _push_scope(
         # pylint: disable=cyclic-import
         from qiskit.circuit.controlflow.builder import ControlFlowBuilderBlock
 
+        # Chain resource requests so things like registers added to inner scopes via conditions are
+        # requested in the outer scope as well.
+        if self._control_flow_scopes:
```

Comment:
This seems like the biggest behavior change (and consequently biggest risk of regression) in this PR as far as I can tell. Can you reply back with a bit more context about why this wasn't needed for bits themselves previously? I see something similar is currently done in `_append`.

Reply:
To be honest, it probably wouldn't have hurt to order things like this originally.  I didn't do it at first because it wasn't _technically_ necessary and I was vaguely concerned that I'd get it wrong and introduce a quadratic complexity in the depth of the scopes.  With only bits, it worked fine because the bits in the inner scope got passed to the outside when the inner scope called `append` on the circuit instance in its `__exit__`.  Your monolithic structure would make this easier.

Reply:
There's little regression risk because resource tracking of bits is very heavily tested, and this is an additive change - most likely the only effect is that some scopes learn about some resources sooner than they otherwise would.  For example, in:
```python
qc = QuantumCircuit(2, 2)
with qc.for_loop(range(2)):
    with qc.for_loop(range(2)):
        qc.h(0).c_if(qc.clbits[0])
```
the outer for loop will now learn that it needs `qc.clbits[0]` when the `c_if` happens, whereas before it would only learn about it when the inner for loop called `append` during its `__exit__` phase.